arbor/communication/dry_run_context.cpp

```diff
@@ -26,7 +24,7 @@ struct dry_run_context_impl {
         count_type local_size = local_spikes.size();
 
         std::vector<spike> gathered_spikes;
-        gathered_spikes.reserve(local_size*num_ranks_);
+        gathered_spikes.reserve(std::size_t{local_size}*num_ranks_);
```

Comment:
We shouldn't change the (effective) maximum size of the gathered_spikes vector without also changing the range of the indices in the loop 35-37.

We ultimately make a partition of the indices into the variable `partition` which has `count_type` as its value type, and this vector is used to construct the `gathered_vector` return type, which also expects things in terms of `count_type`. I believe the correct fix is not to cast `local_size` or use `std::size_t` here but instead
1. Add an assertion which confirms that the `local_size*num_ranks_` doesn't wrap in `count_type`, and
2. Uses `std::size_t{local_size*num_ranks_}` to keep CodeQL happy or else we tell CodeQL that this is not a bug (my preference).